### PR TITLE
Moved count gen to template

### PR DESCRIPTION
**Pull request by dogversioning**
This PR makes the following changes:
- Reorg of counts generator to leverage BaseTableBuilder
- Updates CLI to dynamically build counts tables

### Checklist
- [X] Consider if documentation (like in `docs/`) needs to be updated
- [X] Consider if tests should be added
- [X] Run pylint if you're making changes beyond adding studies
- [X] Update template repo if there are changes to study configuration

### CODE REVIEW

File: cumulus_library/studies/core/count_core.sql
Comment (dogversioning) [inline cumulus_library/studies/core/count_core.sql:1]: 99% of the changes here are indentations just to support the addition of parenthesis after the start of the CTAS statement

File: cumulus_library/studies/core/manifest.toml [hunk 1]
```diff
@@ -23,7 +23,11 @@ file_names = [
     "observation_vital_signs.sql",
     "medication_request.sql",
     "study_period.sql",
-    "count_core.sql"
+]
+
+[count_table_builder_config]
```
Comment (mikix) [inline cumulus_library/studies/core/manifest.toml:28]: nit: you named the class `CountsBuilder` - maybe `counts_builder_config` would be a good symmetry? But no opinion myself

File: tests/test_cli.py [hunk 2]
```diff
@@ -71,6 +71,22 @@ def test_cli_path_mapping(
...
+            cli_args=[
+                "build",
+                "-t",
+                "study_python_counts_valid",
+                "-s" "./tests/test_data",
+                "--database",
+                "test",
+            ]
+        )
+        builder.cursor.execute.assert_called()
```
Comment (mikix) [inline tests/test_cli.py:87]: might be worth testing that it actually did anything interesting too (like make an expected output file)

**Review by mikix — APPROVED** [review]

File: cumulus_library/schema/counts.py [hunk 3]
```diff
@@ -1,128 +1,98 @@
...
+
+
+class CountsBuilder(BaseTableBuilder):
+    """Extends BaseTableBuilder for counts-related use cases"""
+
+    def __init__(self, study_prefix: str = None):
+        if study_prefix is None:
+            # This slightly wonky approach will give us the path of the
+            # directory of a class extending the CountsBuilder
+            study_path = f"{Path(sys.modules[self.__module__].__file__).parent}"
```
[the remark this answers is not in the document]
Comment (dogversioning) [reply cumulus_library/schema/counts.py:20]: good catch, that was a carryover

File: cumulus_library/schema/counts.py [hunk 4]
```diff
@@ -1,128 +1,98 @@
...
+        self, table_name: str, source_table: str, table_cols: list, **kwargs
+    ) -> str:
+        """Wrapper method for generating a counts table from a template"""
+        if not table_name or not source_table or not table_cols:
+            raise CountsBuilderError(
+                "count_query missing required arguments. " f"output table: {table_name}"
+            )
+        for key in kwargs:
+            if key not in ["min_subject", "where_clauses", "cnt_encounter"]:
+                raise CountsBuilderError(f"count_query recieved unexpected key: {key}")
```
[the remark this answers is not in the document]
Comment (dogversioning) [reply cumulus_library/schema/counts.py:64]: there's a particular logic where, depending on the entry point you're coming in from, not all of these are going to be required, which i think is author friendly, so i'm ok just specifying the list of optional params here explicitly.

File: cumulus_library/studies/core/manifest.toml (hunk 1 above)
Comment (dogversioning) [reply cumulus_library/studies/core/manifest.toml:28]: good call

File: tests/test_cli.py (hunk 2 above)
Comment (dogversioning) [reply tests/test_cli.py:87]: yeah - i sort of treated this as a second class test, since it is just testing a different entrypoint into a method that's tested elsewhere, but lemme try and improve this.

Comment (mikix) [reply tests/test_cli.py:87]: I hear ya, and you don't necessarily want to go overboard on an end-to-end unit test. But I like having all the glue between code tested in this way. Which you did! I would just throw a couple more checks in and call it a day.

Comment (dogversioning) [reply tests/test_cli.py:87]: the hard part about this is that the output of all this stuff is 'we executed some queries against athena' - if we had a mock, we could check that data existed.

Comment (mikix) [reply tests/test_cli.py:87]: Ah yeah - I live in the privileged ETL world, where I have files on disk.

If it's hard, don't stress about it, just a thought.

Comment (dogversioning) [reply tests/test_cli.py:92]: I did get in here and at least checked to make sure a sensible query exists in the call arg, which did root out an edge case bug, so it was worth doing.

I was thinking about this a little in the background, and I may open up a ticket with BCH IT to try and enable this approach for regression: https://github.com/smart-on-fhir/cumulus-library/issues/110

File: cumulus_library/study_parser.py [hunk 5]
```diff
@@ -233,19 +241,24 @@ def _load_and_execute_builder(
...
                 f"Error loading {self._study_path}{filename}\n"
-                "Custom builders must extend the BaseTableBuilder "
-                "class exactly once per module."
+                "Custom builders must extend the BaseTableBuilder class."
             )
 
+        # Remove instances of intermediate classes, if present
+        for classref in table_builder_subclasses:
+            if classref.__name__ == "CountsBuilder":
+                table_builder_subclasses.remove(classref)
```
[the remark this answers is not in the document]
Comment (dogversioning) [reply cumulus_library/study_parser.py:260]: yeah i had that same concern - the lambda approach definitely feels safer.
